### ISSUE TITLE
attaching ddt to a set of processes

### ISSUE THREAD

**Issue by antonl321** [LLNL/STAT#45]
I have tried to attach ddt subset of processes and I get the following

`fork exec DDT ['/usr/local/apps/forge/21.1.1/bin/ddt', '--attach-mpi', '233863', '--subset', '0,2,34,37,12,48,58,30', './t3.x']
The specified process was not found.`

I suspect that ddt has to be launch on head node whre srun launch.

Further exploration with attaching ddt to an test code revealed that --attach-mpi is not working with srun (at leat on my system) even on the head node.

 But I can attach with `--attach=[host1:]pid1,[host2:]pid2...` option.
Would it be easy to implement this attaching method in STAT?

**Comment by lee218llnl** [comment]
For the example you gave you attached to "aa3-2035:312897" and I'm guessing you ran that from a login node or launch node. Are you able to ssh to aa3-2035 and launch the STAT GUI there? Once on aa3-2035, you should be able to just attach to the PID, 312897, in your case. Either way, I would then suggest trying to copy the exact ddt command line and all its args and run it on aa3-2035 as a sanity check. Sometimes there are multiple srun processes and you need to be sure to attach to the appropriate one. The forge version should be fine as they have supported this for many years now.

**Comment by lee218llnl** [comment]
As noted in the mention, I put in another PR in spack. The latest gobject introspection did not get the fix/update that we were expecting, so we need to enforce the libffi version for the latest gobject-introspection 1.72.0. Thanks for reporting this and for your patience. I hope the explicit libffi dependence will suffice in the meantime.

**Comment by lee218llnl** [comment]
sorry for the incorrect reference to this issue, I meant to post it to issue 44.

**Comment by antonl321** [comment]
I have launched the job with srun from the login node. In this case the top srun demon (job slurm demon)  runs on the login node and ddt detects the correct srun process to connect. (well on my system ddt fails with the error I have described above).

When a job is submited with sbatch the job slurm deamon runs on the head node of the node allocation.
So, I think that stat makes the fork on the login node instead on the job head node and therefore it cannot find the job slurm demon.

**Comment by lee218llnl** [comment]
I just pushed a change to the develop branch that adds a "DDT serial" button that implements your suggestion (https://github.com/LLNL/STAT/commit/f3827a3c596d5300e35b38b3eb5392867ccdbf7c). It appears to work for me. Can you test it out?

**Comment by antonl321** [comment]
The new attachment works, but I don't get it to what group of nodes is attaching.
In my list I have equivalence classes with 1 node. I have cliked on the middle button for that class but ddt still attaches to a large subset of nodes.
How id this meant to work?

**Comment by lee218llnl** [comment]
By default, STAT will choose a single representative (Rep), the lowest MPI rank) for every equivalence class listed in the Equivalence Classes window. If you don't want a representative for a given equivalence class, please click on the "None' radio button in the EQ Classes window. The "All" should select all ranks listed on that line. Could you perhaps send screenshots of you eq class window before clicking the subset attach and then send the ddt subset attach command that STAT issues (should be printed to the command line).

**Comment by antonl321** [comment]
Ok, I got you. I was confused about how the equivalence classes a are used by ddt. It makes sense now! Thanks.

I think that "DDT serial" is not the best name because this is used for a parallel apps and it might confuse some users. But I don't have a better idea. Essentially the second attach method goes directly to the app processes. I don't see a short name for this.

**Comment by lee218llnl** [comment]
I just pushed another change to the STAT develop branch. I renamed both buttons, so now we have "DDT bulk attach" and "DDT host:PID attach". It's a bit wordy for buttons, but hopefully clarifies the difference. I also updated the documentation about this. Let me know if this looks good to you.

**Comment by antonl321** [comment]
I looks good. Thanks! Closing this issue here.
